Improved heading detection by adding isolated bold case

hnc-leebd opened the pull request:
For the dataloader, I added a case that detects titles with the same font name, a large font size, and bold style, which are occasionally seen in papers.

Since bold style already receives bonus points for properness, I added logic to determine if a short, single-line text node is isolated and assign bonus points.

This logic requires determining the bounding box, which involves expanding the overall test code.

- Improved heading detection by adding isolated bold case.
- Update NodeUtilsTest to now also consider the bounding box of each text node.
- Add AssertJ as a test-scoped dependency so fluent assertions are available.

bundolee committed:
Add AssertJ as a test-scoped dependency so fluent assertions are available

bundolee committed:
Update NodeUtilsTest to now also consider the bounding box of each text node.

MaximPlusov commented on src/main/java/org/verapdf/wcag/algorithms/semanticalgorithms/utils/NodeUtils.java:
```diff
@@ -99,6 +102,22 @@ public static double headingProbability(SemanticTextNode textNode, INode neighbo
+			} else {
+				probability -= HEADING_PROBABILITY_FAR_FROM_NEIGHBOR_BOOST / 2;
+			}
+			if (isSingleLine(textNode) && isShorterThanNeighbor(textNode, neighborTextNode)) {
```
`SemanticTextNode` has method `getLinesNumber`

MaximPlusov commented on src/main/java/org/verapdf/wcag/algorithms/semanticalgorithms/utils/NodeUtils.java:
```diff
@@ -99,6 +102,22 @@ public static double headingProbability(SemanticTextNode textNode, INode neighbo
+			if (isSingleLine(textNode) && isShorterThanNeighbor(textNode, neighborTextNode)) {
+				probability += HEADING_PROBABILITY_SINGLE_LINE_BOOST;
+			} else {
+				probability -= HEADING_PROBABILITY_SINGLE_LINE_BOOST;
```
I'm not sure, that we should reduce probability here. We already use `getLinesNumberHeadingProbability`

hnc-leebd replied:
If getLinesNumberHeadingProbability already exists, this is redundant. Thanks for the review!

MaximPlusov commented on src/main/java/org/verapdf/wcag/algorithms/semanticalgorithms/utils/NodeUtils.java:
```diff
@@ -185,6 +204,46 @@ public static boolean hasSimilarBackgroundColor(double[] firstColor, double[] se
+		return 0.0;
+	}
+
+	private static boolean overlapsVertically(BoundingBox first, BoundingBox second) {
```
Could you please move this method to `BoundingBox`

hnc-leebd replied:
Thank you so much for the great review!

MaximPlusov commented on src/main/java/org/verapdf/wcag/algorithms/semanticalgorithms/utils/NodeUtils.java:
```diff
@@ -99,6 +102,22 @@ public static double headingProbability(SemanticTextNode textNode, INode neighbo
 		}
+
+		if (StaticContainers.isDataLoader()) {
+			if (isFarFromNeighbor(textNode, neighborTextNode, textNode.getFontSize())) {
+				probability += HEADING_PROBABILITY_FAR_FROM_NEIGHBOR_BOOST;
+			} else if (isFarFromNeighbor(textNode, neighborTextNode, textNode.getFontSize() / 2)) {
```
I recommend to use `.getBoundingBox().getHeight()` intead of `.getFontSize()`

hnc-leebd replied:
Thanks for the recommendation!

MaximPlusov commented on src/main/java/org/verapdf/wcag/algorithms/semanticalgorithms/utils/NodeUtils.java:
```diff
@@ -99,6 +102,22 @@ public static double headingProbability(SemanticTextNode textNode, INode neighbo
+			} else if (isFarFromNeighbor(textNode, neighborTextNode, textNode.getFontSize() / 2)) {
+				probability += HEADING_PROBABILITY_FAR_FROM_NEIGHBOR_BOOST / 2;
+			} else {
+				probability -= HEADING_PROBABILITY_FAR_FROM_NEIGHBOR_BOOST / 2;
```
I'm not sure, that we should reduce probability here.

hnc-leebd replied:
I agree. I'm not sure it's absolutely necessary.

hnc-leebd reviewed (commented):
I will push the revised commit.

bundolee committed:
Improved heading detection by adding isolated bold case.

hnc-leebd commented:
@MaximPlusov 
I squashed the third commit and the new commit.
I've incorporated all of the feedback you gave me.
Thank you so much for the detailed review!
